### PR TITLE
Improve logging for UI

### PR DESCRIPTION
**Pull request by elyezer**
Now every time a locator is selected a debug entry mentioning the
locator key and value will be created.

After the changes:

```
2015-03-24 16:55:07 - robottelo.ui.locators - DEBUG - Accessing locator "login.username" by id: "login_login"
2015-03-24 16:55:07 - robottelo.ui.locators - DEBUG - Accessing locator "login.username" by id: "login_login"
2015-03-24 16:55:08 - robottelo.ui.locators - DEBUG - Accessing locator "login.password" by id: "login_password"
2015-03-24 16:55:08 - robottelo.ui.locators - DEBUG - Accessing locator "submit" by name: "commit"
2015-03-24 16:55:13 - robottelo.ui.locators - DEBUG - Accessing locator "notif.error" by xpath: "//div[contains(@class, 'jnotify-notification-error')]"
2015-03-24 16:55:13 - robottelo.ui.locators - DEBUG - Accessing locator "menu.administer" by xpath: "//div[contains(@style,'static') or contains(@style, 'fixed')]//a[@id='administer_menu']"
2015-03-24 16:55:13 - robottelo.ui.locators - DEBUG - Accessing locator "menu.users" by xpath: "//div[contains(@style,'static') or contains(@style, 'fixed')]//a[@id='menu_item_users']"
2015-03-24 16:55:15 - robottelo.ui.locators - DEBUG - Accessing locator "users.new" by xpath: "//a[contains(@href, '/users/new')]"
2015-03-24 16:55:18 - robottelo.ui.locators - DEBUG - Accessing locator "users.new" by xpath: "//a[contains(@href, '/users/new')]"
2015-03-24 16:55:19 - robottelo.ui.locators - DEBUG - Accessing locator "users.username" by id: "user_login"
2015-03-24 16:55:21 - robottelo.ui.locators - DEBUG - Accessing locator "users.username" by id: "user_login"
2015-03-24 16:55:22 - robottelo.ui.locators - DEBUG - Accessing locator "users.firstname" by id: "user_firstname"
2015-03-24 16:55:23 - robottelo.ui.locators - DEBUG - Accessing locator "users.lastname" by id: "user_lastname"
2015-03-24 16:55:23 - robottelo.ui.locators - DEBUG - Accessing locator "users.authorized_by" by id: "user_auth_source_id"
2015-03-24 16:55:23 - robottelo.ui.locators - DEBUG - Accessing locator "users.authorized_by" by id: "user_auth_source_id"
2015-03-24 16:55:24 - robottelo.ui.locators - DEBUG - Accessing locator "users.email" by id: "user_mail"
2015-03-24 16:55:25 - robottelo.ui.locators - DEBUG - Accessing locator "users.email" by id: "user_mail"
2015-03-24 16:55:26 - robottelo.ui.locators - DEBUG - Accessing locator "users.password" by id: "user_password"
2015-03-24 16:55:26 - robottelo.ui.locators - DEBUG - Accessing locator "users.password" by id: "user_password"
2015-03-24 16:55:27 - robottelo.ui.locators - DEBUG - Accessing locator "users.password_confirmation" by id: "user_password_confirmation"
2015-03-24 16:55:27 - robottelo.ui.locators - DEBUG - Accessing locator "users.password_confirmation" by id: "user_password_confirmation"
2015-03-24 16:55:28 - robottelo.ui.locators - DEBUG - Accessing locator "submit" by name: "commit"
2015-03-24 16:55:33 - robottelo.ui.locators - DEBUG - Accessing locator "menu.administer" by xpath: "//div[contains(@style,'static') or contains(@style, 'fixed')]//a[@id='administer_menu']"
2015-03-24 16:55:33 - robottelo.ui.locators - DEBUG - Accessing locator "menu.users" by xpath: "//div[contains(@style,'static') or contains(@style, 'fixed')]//a[@id='menu_item_users']"
2015-03-24 16:55:38 - robottelo.ui.locators - DEBUG - Accessing locator "users.user" by xpath: "//a[contains(., '%s')]"
2015-03-24 16:55:38 - robottelo.ui.locators - DEBUG - Accessing locator "search" by id: "search"
2015-03-24 16:55:38 - robottelo.ui.locators - DEBUG - Accessing locator "search_button" by xpath: "//button[contains(@type,'submit')]"
2015-03-24 16:55:44 - robottelo.ui.locators - DEBUG - Accessing locator "login.gravatar" by xpath: "//img[contains(@class, 'avatar')]"
2015-03-24 16:55:44 - robottelo.ui.locators - DEBUG - Accessing locator "menu.account" by xpath: "//a[@id='account_menu']"
2015-03-24 16:55:44 - robottelo.ui.locators - DEBUG - Accessing locator "menu.sign_out" by xpath: "//a[@id='menu_item_logout']"
```

### CODE REVIEW

File: robottelo/ui/locators.py [hunk 1]
```diff
@@ -1,14 +1,35 @@
...
+
+class LocatorDict(collections.Mapping):
+    """This class will log every time an item is selected"""
+    def __init__(self, *args, **kwargs):
+        self.store = dict(*args, **kwargs)
+
+    def __getitem__(self, key):
+        item = self.store[key]
+        LOGGER.debug(
+            'Accessing locator "%s" by %s: "%s"', key, item[0], item[1]
```
Comment (JacobCallahan) [inline robottelo/ui/locators.py]: would it be better to use .format() here?

Comment (Ichimonji10) [inline robottelo/ui/locators.py]: Nope! When you call `debug`, a `logging.LogRecord` object is created. This object stores both the logging formatting string and the objects that will be typecast and injected into the string. Why is this done? See [here](https://docs.python.org/3/library/logging.html#logging.LogRecord):

> This functionality can be used to inject your own values into a LogRecord at creation time.

In Python 3, the `logging.Formatter` class accepts a `style` argument. This argument is `'%'` by default, but can be set to `'{'` or `'$'` instead. Unfortunately, this functionality is not available in Python 2.

File: robottelo/ui/locators.py [hunk 2]
```diff
@@ -1,14 +1,35 @@
...
 
 
-menu_locators = {
+LOGGER = logging.getLogger(__name__)
+
+
+class LocatorDict(collections.Mapping):
+    """This class will log every time an item is selected"""
+    def __init__(self, *args, **kwargs):
+        self.store = dict(*args, **kwargs)
```
Comment (Ichimonji10) [inline robottelo/ui/locators.py]: Would it be simpler to just accept a single argument? For example:

```python
def __init__(self, dictionary):
    self.store = dictionary
```

Or are there techniques for dictionary creation that necessitate that `*args` and `**kwargs` be accepted instead?

**Comment by Ichimonji10** [comment]
ACK pending answer to comment.

**Comment by JacobCallahan** [comment]
ACK

File: robottelo/ui/locators.py (hunk 1 above)
Comment (JacobCallahan) [inline robottelo/ui/locators.py]: interesting. good to know! :)

File: robottelo/ui/locators.py (hunk 2 above)
Comment (elyezer) [inline robottelo/ui/locators.py]: Basically I want to support dict creation like below:

```python
>>> dict({'a': 'b'})
{'a': 'b'}
>>> dict(a='b')
{'a': 'b'}
```

**Comment by elyezer** [comment]
This is not the ideal solution but is an easier way to improve de logging without the need of a bigger refactor.

File: robottelo/ui/locators.py (hunk 2 above)
Comment (Ichimonji10) [inline robottelo/ui/locators.py]: OK. That answers my question. Can you add a comment into the code or a brief explanation into the docstring?

```python
def __init__(self, *args, **kwargs):
    # >>> dict({'a': 'b'})
    # {'a': 'b'}
    # >>> dict(a='b')
    # {'a': 'b'}
    self.store = dict(*args, **kwargs)
```

**Comment by Ichimonji10** [comment]
APT

**Pull request merged**